File: xai/src/milliampere_xai/rendering/_shap_bars.py
"""SHAP bar chart window showing per-observation feature contributions."""

from __future__ import annotations

from typing import Any

import numpy as np
import pygame
from milliampere_dp.rendering import Color
from milliampere_dp.vessel import MAX_THRUSTER_RPM

from milliampere_xai.rendering._window import Window


class ShapRender(Window):
    """SHAP bar chart window showing per-observation feature contributions."""

    def __init__(
        self,
        screen,
        window_pos,
        legend_items,
        title="SHAP-value",
        window_width=450,
        window_height=450,
        renderer=None,
    ):
        super().__init__(
            screen, window_pos, title, window_width, window_height, renderer=renderer
        )

        # fonts
        self.axis_font = pygame.font.SysFont("DejaVu Sans", 13)
        self.label_font = pygame.font.SysFont("DejaVu Sans", 12)

        self.window_padding = 40  # pixels
        self.num_bars = 4  # 14 # number of features
        self.num_bar_seg = 4  # number of outputs
        self.max_shap_value = 4
        self.increment = 1

        self.bar_gap = 25  # pixels
        self.label_offset = 10  # pixels
        self.label_surface: list[Any] = [None] * 14
        self.label_rect: list[Any] = [None] * 14
        self.bars_py: list[Any] = [None] * self.num_bars

        self.explain_mode = -1
        self.RPM_true = True

        self.feature_names = [
            "\u1e8d\u1d47\u209c [m]",
            "\u1ef3\u1d47\u209c [m]",
            "\u03c8\u0303\u209c[\u00b0]",
            "\u00fb\u209c [m/s]",
            "\u1e7d\u0302\u209c [m/s]",
            "r\u0302\u209c [\u00b0/s]",
            "n_{\u2093\u2081,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u1d67\u2081,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u2093\u2082,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u1d67\u2082,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u2093\u2083,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u1d67\u2083,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u2093\u2084,d,\u209c\u208b\u2081} [RPM]",
            "n_{\u1d67\u2084,d,\u209c\u208b\u2081} [RPM]",
        ]
        self.feature_names_full = self.feature_names[:]  # <- new

        # colors
        self.colors = [
            Color.TABLEAU_BLUE.value,
            Color.TABLEAU_ORANGE.value,
            Color.TABLEAU_GREEN.value,
            Color.TABLEAU_RED.value,
            Color.GREEN.value,
            Color.RED.value,
        ]

        # making labels
        for i in range(14):
            self.label_surface[i] = self.label_font.render(
                self.feature_names[i], self.antialias, Color.BLACK.value
            )
            self.label_rect[i] = self.label_surface[i].get_rect()

        # defining axis padding width and bar allocation width
        self.axis_padding_width = (
            max([self.label_rect[i].width for i in range(14)]) + self.label_offset
        )
        self.bar_allocation_width = (
            self.window_width - self.window_padding - self.axis_padding_width
        )

        self.label_surface: list[Any] = [None] * self.num_bars
        self.label_rect: list[Any] = [None] * self.num_bars

        # making labels
        for i in range(self.num_bars):
            self.label_surface[i] = self.label_font.render(
                self.feature_names[i], self.antialias, Color.BLACK.value
            )
            self.label_rect[i] = self.label_surface[i].get_rect()

        # making bottom text
        bottom_text = "Sum of SHAP based thrust components"
        self.bottom_text_surface = self.axis_font.render(
            bottom_text, self.antialias, Color.BLACK.value
        )
        self.bottom_text_rect = self.bottom_text_surface.get_rect()
        self.bottom_text_rect.midbottom = (
            (self.window_width + self.axis_padding_width) / 2,
            self.window_height - self.window_padding / 2,
        )

        # making ticks
        self.num_ticks = int(self.max_shap_value / self.increment) + 1
        self.ticks_px: list[Any] = [None] * self.num_ticks
        self.tick_label: list[Any] = [None] * self.num_ticks
        self.tick_rect: list[Any] = [None] * self.num_ticks
        for tick in range(self.num_ticks):
            tick_value = tick * self.increment
            self.ticks_px[tick] = (
                self.window_padding / 2
                + self.axis_padding_width
                + self._shap_value_2_pixels(tick_value)
            )
            self.tick_label[tick] = self.label_font.render(
                f"{tick_value * MAX_THRUSTER_RPM}", self.antialias, Color.BLACK.value
            )
            self.tick_rect[tick] = self.tick_label[tick].get_rect()
            self.tick_rect[tick].midbottom = (
                self.ticks_px[tick],
                self.bottom_text_rect.midtop[1] - 5,
            )

        # defining axis padding height
        self.axis_padding_height = (
            self.bottom_text_rect.height
            + 5
            + self.tick_rect[0].height
            + self.label_offset
        )

        # defining rest of bar parameters
        self.bar_allocation_hight = (
            self.window_height - self.window_padding - self.axis_padding_height
        ) / self.num_bars
        self.bar_height = self.bar_allocation_hight - self.bar_gap

        # defining axis positions
        self.origo = (
            self.window_padding / 2 + self.axis_padding_width,
            self.window_height - self.window_padding / 2 - self.axis_padding_height,
        )
        self.x_axis_end = (
            self.window_width - self.window_padding / 2,
            self.window_height - self.window_padding / 2 - self.axis_padding_height,
        )
        self.y_axis_end = (
            self.window_padding / 2 + self.axis_padding_width,
            self.window_padding / 2,
        )

        # defining position for labels
        for i in range(self.num_bars):
            self.bars_py[i] = self.window_padding / 2 + i * self.bar_allocation_hight
            self.label_rect[i].topright = (
                self.window_padding / 2 + self.axis_padding_width - self.label_offset,
                self.bars_py[i] + (self.bar_height - self.label_rect[i].height) / 2,
            )

        self.legend_items = [
            (self.colors[i], legend) for i, legend in enumerate(legend_items)
        ]
        self.legend_surface, self.box_pos = self.create_legend(
            self.legend_items, self.label_font, margin=self.window_padding / 2
        )

    # public functions ############################

    def render(self, shap_values, shap_values3, base_value, action_low, action_high):
        # init
        self.render_surface()

        # SHAP render
        self._draw_bars2(shap_values, shap_values3, base_value, action_low, action_high)
        self._draw_axis()
        self.draw_legend(self.legend_surface, self.box_pos)

        # final render
        self.render_surface_title()
        self.render_window()

    # private functions ############################

    def _shap_value_2_pixels(self, shap_value):
        return int(shap_value / self.max_shap_value * self.bar_allocation_width)

    def _draw_bars2(
        self, shap_values_action, shap_values_vf, base_value, action_low, action_high
    ):
        """
        Draw four horizontal bars:
        \u2013 the three most influential individual features
        \u2013 one aggregated \u201cOthers\u201d bar
        """

        # shap_values_action (8x14)
        # shap_values_vf (14)

        sv_action = np.array(shap_values_action)
        sv_base = np.array(base_value)
        act_low = np.array(action_low)
        act_high = np.array(action_high)

        sv_all_actions = np.clip(sv_action.sum(axis=1) + sv_base, act_low, act_high)

        pos_totalt = np.where(sv_action > 0, sv_action, 0).sum(axis=1) + np.where(
            sv_base > 0, sv_base, 0
        )
        neg_totalt = np.where(sv_action < 0, sv_action, 0).sum(axis=1) + np.where(
            sv_base < 0, sv_base, 0
        )

        denominators = np.where(act_high == 1, pos_totalt, neg_totalt)

        c = np.zeros_like(sv_all_actions, dtype=float)

        np.divide(
            sv_all_actions, denominators, out=c, where=denominators != 0
        )  # Avoid division by zero

        # sv_action: shape (8, 14)
        # c:          shape (8,)
        # We\u2019ll treat (0,1), (2,3), (4,5), (6,7) as your (x,y) pairs.

        # 1) First, reshape into \u201cpairs\u201d:
        num_actions, num_feats = sv_action.shape
        assert num_actions % 2 == 0, "Need an even number of action\u2010rows"
        sv_pairs = sv_action.reshape(-1, 2, num_feats)  # shape (4, 2, 14)
        c_pairs = c.reshape(-1, 2)  # shape (4, 2)

        # 2) Scale each x\u2010row by its c and each y\u2010row by its c:
        #    we broadcast c_pairs (4,2) over the features\u2010axis
        scaled = sv_pairs * c_pairs[:, :, None]  # shape (4, 2, 14)

        # 3) Square + sum over the x/y axis, then sqrt \u2192 Euclidean distance
        # sq       = scaled ** 2                           # shape (4, 2, 14)
        # sum_sq   = sq.sum(axis=1)                        # shape (4, 14)
        # dists    = np.sqrt(sum_sq)                       # shape (4, 14)
        dists = np.hypot(scaled[:, 0, :], scaled[:, 1, :])

        # 4) Clip each distance between 0 and 1
        dists_clipped = np.clip(dists, 0, 1)  # shape (4, 14)
        sumed_features = dists_clipped.sum(axis=0)

        top3_idx = np.argpartition(sumed_features, -3)[-3:]  # get any order of top3
        top3_idx = top3_idx[
            np.argsort(-sumed_features[top3_idx])
        ]  # then sort those three
        top3_idx = list(top3_idx)
        # print(f"top3_idx:{top3_idx}")

        others_idx = [
            i for i in range(len(self.feature_names_full)) if i not in top3_idx
        ]

        draw_order = top3_idx + [-1]  # \u20131 will mean \u201cOTHERS\u201d

        # --------------------------------------------------
        # 3. loop over the four bars to draw
        # --------------------------------------------------
        for bar_no, feat_idx in enumerate(draw_order):
            # ---------- build the 4 coloured segments ----------
            seg_width_px = []  # length-4 list of pixel widths
            for seg in range(self.num_bar_seg):
                if feat_idx == -1:  # -------- OTHERS ------------
                    # sum this segment's contribution over *all* residual features
                    seg_val = 0.0
                    for jj in others_idx:
                        seg_val += dists_clipped[seg, jj]
                else:  # -------- single feature ----
                    seg_val = dists_clipped[seg, feat_idx]
                seg_width_px.append(self._shap_value_2_pixels(seg_val))

            # ---------- actually draw the bar ----------
            x0 = self.window_padding / 2 + self.axis_padding_width
            y0 = self.bars_py[bar_no]
            for seg, w in enumerate(seg_width_px):
                self._renderer.draw_rect(
                    self.surface, self.colors[seg], (x0, y0, w, self.bar_height)
                )
                x0 += w  # next segment starts where the last finished

            # ---------- draw the label ----------
            label = "Others" if feat_idx == -1 else self.feature_names_full[feat_idx]
            self.label_surface[bar_no] = self.label_font.render(
                label, self.antialias, Color.BLACK.value
            )
            self.label_rect[bar_no] = self.label_surface[bar_no].get_rect()
            self.label_rect[bar_no].topright = (
                self.window_padding / 2 + self.axis_padding_width - self.label_offset,
                y0 + (self.bar_height - self.label_rect[bar_no].height) / 2,
            )
            self.surface.blit(self.label_surface[bar_no], self.label_rect[bar_no])

    def _draw_axis(self):
        self._renderer.draw_line(
            self.surface, Color.BLACK.value, self.origo, self.x_axis_end, width=2
        )  # x-axis
        self._renderer.draw_line(
            self.surface, Color.BLACK.value, self.origo, self.y_axis_end, width=2
        )  # y-axis

        for tick in range(self.num_ticks):
            self._renderer.draw_line(
                self.surface,
                Color.BLACK.value,
                (self.ticks_px[tick], self.origo[1] - 5),
                (self.ticks_px[tick], self.origo[1] + 5),
                width=2,
            )
            self.surface.blit(self.tick_label[tick], self.tick_rect[tick])

        self.surface.blit(self.bottom_text_surface, self.bottom_text_rect)

File: web/src/views/shap_bars.ts
/**
 * SHAP horizontal bar chart visualization.
 * Faithful port of milliampere_xai/rendering/_shap_bars.py (ShapRender).
 *
 * Shows the top-3 most influential features + aggregated "Others" as four
 * horizontal stacked bars.  Each bar has 4 coloured segments representing
 * the 4 thruster pairs.
 */

import { Colors } from "../colors.ts";
import type { Renderer } from "../renderer.ts";
import type { RenderFrame } from "../types.ts";

// ── Constants matching Python ShapRender.__init__ ──────────────────────

const MAX_THRUSTER_RPM = 900;

const FEATURE_NAMES = [
  "x̃ᵇₜ [m]",
  "ỹᵇₜ [m]",
  "ψ̃ₜ[°]",
  "ûₜ [m/s]",
  "v̂ₜ [m/s]",
  "r̂ₜ [°/s]",
  "n_{ₓ₁,d,ₜ₋₁} [RPM]",
  "n_{ᵧ₁,d,ₜ₋₁} [RPM]",
  "n_{ₓ₂,d,ₜ₋₁} [RPM]",
  "n_{ᵧ₂,d,ₜ₋₁} [RPM]",
  "n_{ₓ₃,d,ₜ₋₁} [RPM]",
  "n_{ᵧ₃,d,ₜ₋₁} [RPM]",
  "n_{ₓ₄,d,ₜ₋₁} [RPM]",
  "n_{ᵧ₄,d,ₜ₋₁} [RPM]",
];

const BAR_COLORS = [
  Colors.TABLEAU_BLUE,
  Colors.TABLEAU_ORANGE,
  Colors.TABLEAU_GREEN,
  Colors.TABLEAU_RED,
];

const LEGEND_ITEMS_THRUST = [
  "n_{₁,d,ₜ} [RPM]",
  "n_{₂,d,ₜ} [RPM]",
  "n_{₃,d,ₜ} [RPM]",
  "n_{₄,d,ₜ} [RPM]",
];

const LEGEND_ITEMS_ANGLE = [
  "α_{d₁,ₜ} [°]",
  "α_{d₂,ₜ} [°]",
  "α_{d₃,ₜ} [°]",
  "α_{d₄,ₜ} [°]",
];

const NUM_BARS = 4; // top 3 + Others
const NUM_BAR_SEG = 4; // one segment per thruster pair
const MAX_SHAP_VALUE = 4;
const INCREMENT = 1;
const WINDOW_PADDING = 40;
const BAR_GAP = 25;
const LABEL_OFFSET = 10;
const NUM_FEATURES = 14;

// ── Helpers ────────────────────────────────────────────────────────────

/** Clamp v to [lo, hi]. */
function clamp(v: number, lo: number, hi: number): number {
  return v < lo ? lo : v > hi ? hi : v;
}

/** Convert SHAP magnitude to pixel width within the bar area. */
function shapValueToPixels(value: number, barAllocationWidth: number): number {
  return (value / MAX_SHAP_VALUE) * barAllocationWidth;
}

/**
 * Compute the 4x14 distance matrix and normalisation factor `c`.
 *
 * Mirrors _draw_bars2 (and _find_explanation2) from the Python code:
 *   1. sv_all_actions = clip(sv_action.sum(axis=1) + sv_base, act_low, act_high)
 *   2. c = sv_all_actions / (pos_total or neg_total depending on act_high)
 *   3. reshape (8,14) -> (4,2,14), scale by c, Euclidean distance, clip to [0,1]
 */
function computeDistances(
  svAction: number[][],
  baseValue: number[],
  actionLow: number[],
  actionHigh: number[],
): number[][] {
  const numActions = svAction.length; // 8
  const numFeats = svAction[0].length; // 14

  // sv_all_actions = clip(sv_action.sum(axis=1) + sv_base, act_low, act_high)
  const svAllActions: number[] = new Array(numActions);
  for (let a = 0; a < numActions; a++) {
    let rowSum = 0;
    for (let f = 0; f < numFeats; f++) rowSum += svAction[a][f];
    svAllActions[a] = clamp(rowSum + baseValue[a], actionLow[a], actionHigh[a]);
  }

  // pos/neg totals
  const posTotal: number[] = new Array(numActions);
  const negTotal: number[] = new Array(numActions);
  for (let a = 0; a < numActions; a++) {
    let pos = 0;
    let neg = 0;
    for (let f = 0; f < numFeats; f++) {
      const v = svAction[a][f];
      if (v > 0) pos += v;
      else neg += v;
    }
    pos += baseValue[a] > 0 ? baseValue[a] : 0;
    neg += baseValue[a] < 0 ? baseValue[a] : 0;
    posTotal[a] = pos;
    negTotal[a] = neg;
  }

  // c = sv_all_actions / denominator  (where denominator depends on act_high)
  const c: number[] = new Array(numActions);
  for (let a = 0; a < numActions; a++) {
    const denom = actionHigh[a] === 1 ? posTotal[a] : negTotal[a];
    c[a] = denom !== 0 ? svAllActions[a] / denom : 0;
  }

  // Reshape into pairs (4, 2, 14) and scale
  const numPairs = numActions / 2;
  // dists: shape (4, 14)
  const dists: number[][] = [];
  for (let p = 0; p < numPairs; p++) {
    const row: number[] = new Array(numFeats);
    const xIdx = p * 2;
    const yIdx = p * 2 + 1;
    for (let f = 0; f < numFeats; f++) {
      const scaledX = svAction[xIdx][f] * c[xIdx];
      const scaledY = svAction[yIdx][f] * c[yIdx];
      row[f] = clamp(Math.hypot(scaledX, scaledY), 0, 1);
    }
    dists.push(row);
  }

  return dists;
}

// ── Legend drawing ──────────────────────────────────────────────────────

function drawLegend(
  r: Renderer,
  ox: number,
  oy: number,
  w: number,
  h: number,
  items: string[],
  padding: number,
): void {
  const markerSize = 12;
  const spacing = 5;
  const legendPadding = 10;
  const fontSize = 12;
  const lineHeight = 14;

  const maxTextWidth = Math.max(...items.map((label) => r.measureText(label, fontSize)));
  const boxW = legendPadding * 2 + markerSize + spacing + maxTextWidth;
  const totalContentH =
    items.length * lineHeight + (items.length - 1) * spacing;
  const boxH = legendPadding * 2 + totalContentH;

  // Position: bottom-right with margin, then up by padding_bottom=60
  const margin = padding / 2;
  const paddingBottom = 60;
  const boxX = ox + w - margin - boxW;
  const boxY = oy + h - margin - paddingBottom - boxH;

  // Semi-transparent background
  r.drawRect(boxX, boxY, boxW, boxH, Colors.LEGEND_BOX);

  let yOff = boxY + legendPadding;
  for (let i = 0; i < items.length; i++) {
    const markerY = yOff + (lineHeight - markerSize) / 2;
    r.drawRect(
      boxX + legendPadding,
      markerY,
      markerSize,
      markerSize,
      BAR_COLORS[i],
    );
    r.drawText(
      items[i],
      boxX + legendPadding + markerSize + spacing,
      yOff,
      Colors.BLACK,
      fontSize,
    );
    yOff += lineHeight + spacing;
  }
}

// ── Main render function ───────────────────────────────────────────────

export function renderShapBars(
  r: Renderer,
  frame: RenderFrame,
  ox: number,
  oy: number,
  w: number,
  h: number,
  _mode: "thrust" | "angle",
): void {
  // White background
  r.drawRect(ox, oy, w, h, Colors.WHITE);

  const shap = frame.shap;
  const svAction = shap.shap_values_action; // (8, 14)
  const baseValue = shap.base_vectors; // (8,)
  const actionLow = shap.action_low; // (8,)
  const actionHigh = shap.action_high; // (8,)

  // ── Layout computation (mirrors Python __init__) ───────────────────

  // Approximate axis_padding_width: widest feature label + label_offset
  // Python uses pygame font metrics; we approximate with 12px * ~10 chars
  const axisPaddingWidth = 110 + LABEL_OFFSET;

  const barAllocationWidth = w - WINDOW_PADDING - axisPaddingWidth;

  // Bottom text height approximation (fontSize 13 + tick labels 12 + margins)
  const bottomTextHeight = 16;
  const tickLabelHeight = 14;
  const axisPaddingHeight =
    bottomTextHeight + 5 + tickLabelHeight + LABEL_OFFSET;

  const barAllocationHeight =
    (h - WINDOW_PADDING - axisPaddingHeight) / NUM_BARS;
  const barHeight = barAllocationHeight - BAR_GAP;

  // Origin (bottom-left of bar chart)
  const origoX = ox + WINDOW_PADDING / 2 + axisPaddingWidth;
  const origoY = oy + h - WINDOW_PADDING / 2 - axisPaddingHeight;

  // Top of y-axis
  const yAxisEndY = oy + WINDOW_PADDING / 2;

  // Right end of x-axis
  const xAxisEndX = ox + w - WINDOW_PADDING / 2;

  // ── Compute SHAP distances ────────────────────────────────────────

  const distsClipped = computeDistances(
    svAction,
    baseValue,
    actionLow,
    actionHigh,
  );

  // Sum across 4 thruster pairs for each feature -> (14,)
  const sumedFeatures: number[] = new Array(NUM_FEATURES).fill(0);
  for (let f = 0; f < NUM_FEATURES; f++) {
    for (let p = 0; p < 4; p++) {
      sumedFeatures[f] += distsClipped[p][f];
    }
  }

  // Find top-3 feature indices by total contribution (descending)
  const indices = Array.from({ length: NUM_FEATURES }, (_, i) => i);
  indices.sort((a, b) => sumedFeatures[b] - sumedFeatures[a]);
  const top3Idx = indices.slice(0, 3);

  const othersIdx = indices.slice(3);

  // Draw order: top3 + "Others" sentinel (-1)
  const drawOrder: number[] = [...top3Idx, -1];

  // ── Draw bars ─────────────────────────────────────────────────────

  for (let barNo = 0; barNo < drawOrder.length; barNo++) {
    const featIdx = drawOrder[barNo];
    const barY = oy + WINDOW_PADDING / 2 + barNo * barAllocationHeight;

    // Build 4 segment widths
    const segWidths: number[] = [];
    for (let seg = 0; seg < NUM_BAR_SEG; seg++) {
      let segVal: number;
      if (featIdx === -1) {
        // Others: sum across all residual features
        segVal = 0;
        for (const jj of othersIdx) {
          segVal += distsClipped[seg][jj];
        }
      } else {
        segVal = distsClipped[seg][featIdx];
      }
      segWidths.push(shapValueToPixels(segVal, barAllocationWidth));
    }

    // Draw stacked segments left-to-right from axis origin
    let x0 = origoX;
    for (let seg = 0; seg < NUM_BAR_SEG; seg++) {
      if (segWidths[seg] > 0.5) {
        r.drawRect(x0, barY, segWidths[seg], barHeight, BAR_COLORS[seg]);
      }
      x0 += segWidths[seg];
    }

    // Feature label (right-aligned to left of bar area)
    const label = featIdx === -1 ? "Others" : FEATURE_NAMES[featIdx];
    r.drawText(
      label,
      origoX - LABEL_OFFSET,
      barY + barHeight / 2,
      Colors.BLACK,
      12,
      "right",
      "middle",
    );
  }

  // ── Draw axes ─────────────────────────────────────────────────────

  // X-axis (horizontal)
  r.drawLine(origoX, origoY, xAxisEndX, origoY, Colors.BLACK, 2);
  // Y-axis (vertical)
  r.drawLine(origoX, origoY, origoX, yAxisEndY, Colors.BLACK, 2);

  // Tick marks and labels
  const numTicks = MAX_SHAP_VALUE / INCREMENT + 1;
  // Bottom text position — Python: bottom_text_rect.midbottom at (mid, h - padding/2)
  // tick_rect.midbottom at (tickPx, bottom_text_rect.midtop[1] - 5)
  const bottomTextY = oy + h - WINDOW_PADDING / 2;
  const tickLabelY = bottomTextY - bottomTextHeight - 10;

  for (let tick = 0; tick < numTicks; tick++) {
    const tickValue = tick * INCREMENT;
    const tickPx =
      WINDOW_PADDING / 2 + axisPaddingWidth +
      shapValueToPixels(tickValue, barAllocationWidth);
    const tickScreenX = ox + tickPx;

    // Tick mark
    r.drawLine(
      tickScreenX,
      origoY - 5,
      tickScreenX,
      origoY + 5,
      Colors.BLACK,
      2,
    );

    // Tick label
    const rpmLabel = `${tickValue * MAX_THRUSTER_RPM}`;
    r.drawText(rpmLabel, tickScreenX, tickLabelY, Colors.BLACK, 12, "center");
  }

  // Bottom text
  const bottomTextX = ox + (w + axisPaddingWidth) / 2;
  r.drawText(
    "Sum of SHAP based thrust components",
    bottomTextX,
    bottomTextY - bottomTextHeight / 2,
    Colors.BLACK,
    13,
    "center",
    "middle",
  );

  // ── Legend ─────────────────────────────────────────────────────────

  const legendItems =
    _mode === "thrust" ? LEGEND_ITEMS_THRUST : LEGEND_ITEMS_ANGLE;
  drawLegend(r, ox, oy, w, h, legendItems, WINDOW_PADDING);

  // ── Title (top-right, matching Python render_surface_title) ───────

  const title =
    _mode === "thrust"
      ? "Feature importance using thrust SHAP-values"
      : "SHAP-values azimuth angles";
  r.drawText(title, ox + w - 5, oy + 5, Colors.GRAY, 11, "right");
}
